src/Documentation/SidebarMenu/helper.js

```diff
@@ -0,0 +1,149 @@
+import startCase from 'lodash.startcase'
+
+import sidebar from '../sidebar'
+
+/*
+  We will use this helper to normalize sidebar structure and create
+  all of the resurces we need to prevent future recalculations.
+
+  Target structure example:
+
+  {
+    label: "Add Files or Directories",
+    path: "/doc/get-started/add-files",
+    source: "/static/docs/get-started/add-files.md",
+    prev: "/doc/get-started/configure",
+    next: "/doc/get-started/share-data",
+    children: []
+  }
+*/
+
+const PATH_ROOT = '/doc/'
+const FILE_ROOT = '/static/docs/'
+const FILE_EXTENSION = '.md'
+
+// Inner helpers
+
+function findItem(data, targetPath) {
+  if (data.length) {
+    for (let i = 0; i < data.length; i++) {
+      const { path, children } = data[i]
+
+      if (path === targetPath) {
+        return data[i]
+      } else if (children) {
+        const result = findItem(children, targetPath)
+        if (result) {
+          return result
+        }
+      }
+    }
+  }
+}
+
+function findChildWithSource(item) {
+  return item.source ? item : findChildWithSource(item.children[0])
+}
+
+function findPrevItemWithSource(data, item) {
+  if (item.source) {
+    return item
+  } else if (item.prev) {
+    const prevItem = findItem(data, item.prev)
+
+    return findPrevItemWithSource(data, prevItem)
+  }
+}
+
+function validateRawItem({ slug, source, children }) {
+  const isSourceDisabled = source === false // is source set to 'false'?
+
+  if (!slug) {
+    throw Error("'slug' field is required in objects in sidebar.json")
+  }
+
+  if (isSourceDisabled && (!children || !children.length)) {
+    throw Error(
+      "If you set 'source' to false, you had to add at least one child"
+    )
+  }
+}
+
+// Global cache vars used in normalization
```

Comment:
can we pass them between functions, do we actually need a global prevReference, for example?

Reply:
I have an idea how it can be done, but I'm not sure that it will make code easier to understand, but I will try today.